catroid/src/main/java/org/catrobat/catroid/utils/ImportProjectsFromExternalStorage.java

```diff
@@ -0,0 +1,169 @@
+	private Activity activity;
+
+	public ImportProjectsFromExternalStorage(Activity mActivity) {
+		activity = mActivity;
+	}
+
+	private static File copyProjectsFromExternalToInternalStorage(File srcDir, File dstDir) throws IOException {
+		UniqueNameProvider uniqueNameProvider = new UniqueNameProvider();
+		List<String> scope = new ArrayList<>();
+		if (!srcDir.exists()) {
+			throw new FileNotFoundException("Directory: " + srcDir.getAbsolutePath() + " does not exist.");
+		}
+
+		if (!srcDir.isDirectory()) {
+			throw new IOException(srcDir.getAbsolutePath() + " is not a directory.");
+		}
+		if (!dstDir.exists()) {
+			dstDir.mkdir();
+		}
+
+		if (!dstDir.isDirectory()) {
+			throw new IOException("Cannot create directory: " + dstDir.getAbsolutePath());
+		}
+
+		for (File srcFile : srcDir.listFiles()) {
+			String srcFileName = srcFile.getName();
+			if (srcFile.isDirectory()) {
+				boolean fileExists = false;
+				for (File dstFile : dstDir.listFiles()) {
+					if (dstFile.getName().equals(srcFileName) && dstFile.getParent().endsWith("Programs") && srcFile.getParent().endsWith("Pocket Code") && !srcFileName.equals("backpack")) {
+						fileExists = true;
+						scope.add(srcFileName);
+						break;
+					}
+				}
+				if (fileExists) {
+					String newFileName = uniqueNameProvider.getUniqueName(srcFileName, scope);
+					copyProjectsFromExternalToInternalStorage(srcFile, new File(dstDir, newFileName));
+				} else {
+					copyProjectsFromExternalToInternalStorage(srcFile, new File(dstDir, srcFileName));
+				}
+			} else {
+				transferData(srcFile, new File(dstDir, srcFileName));
+			}
+		}
+
+		return dstDir;
+	}
+
+	public void importOldPocketCodePrograms() {
+		if (FlavoredConstants.DEFAULT_ROOT_DIRECTORY.equals(POCKET_CODE_ROOT_DIRECTORY) && pocketCodeExternalStorageDirectoryExists()) {
+			showImportProjectsDialog(activity);
+		}
+	}
+
+	private boolean pocketCodeExternalStorageDirectoryExists() {
+		return EXTERNAL_STORAGE_ROOT_DIRECTORY.exists();
+	}
+
+	private void showImportProjectsDialog(Activity activity) {
+		AlertDialog.Builder alertDialog = new AlertDialog.Builder(activity);
+		alertDialog.setTitle(R.string.import_dialog_title)
+				.setCancelable(false)
+				.setMessage(R.string.import_dialog_message)
+				.setPositiveButton(R.string.ok, this)
+				.setNeutralButton(R.string.import_dialog_copy_only, this);
+
+		boolean showDialog = PreferenceManager
+				.getDefaultSharedPreferences(activity).getBoolean(SHOW_COPY_PROJECTS_FROM_EXTERNAL_STORAGE_DIALOG, true);
+		if (showDialog) {
+			alertDialog.show();
+		}
+	}
+
+	@Override
+	public void onClick(DialogInterface dialog, int which) {
+		switch (which) {
+			case DialogInterface.BUTTON_POSITIVE:
+				copyProjects();
+				try {
+					StorageOperations.deleteDir(EXTERNAL_STORAGE_ROOT_DIRECTORY);
+				} catch (IOException e) {
+					Log.e(TAG, e.getMessage());
+				}
+				doNotShowDialogAgain();
+				break;
+			case DialogInterface.BUTTON_NEUTRAL:
+				copyProjects();
```

Comment:
since this could take some time, it should not run on the main thread (hearing stories about 100mb+ sized projects out there).
android has this uithread watchdog timer thingy running, and after a rather short time of a busy/frozen ui thread will open this dialog "Pocket code isnt responding" that offers the user to kill the app. (which for obvious reasons would be bad to do whilst copying stuff around)

so this should be in an async task. (of course also applies to above call to copyProjects())

Reply:
should ProgressDialog be shown ?